docs: x/distribution spec (#8928)
cosmos/cosmos-sdk: x/distribution/spec/03_begin_block.md

<!--
order: 3
-->

# Begin Block

> At each `BeginBlock`, the fees received in the previous block are transferred to the distribution `ModuleAccount`, as it's the account the one who keeps track of the flow of coins in (as in this case) and out the module. The fees are also allocated to the proposer, community fund and global pool. When the validator is the proposer of the round, that validator (and their delegators) receives between 1% and 5% of fee rewards, the reserve community tax is then charged, then the remainder is distributed proportionally by voting power to all bonded validators independent of whether they voted (social distribution). Note the social distribution is applied to proposer validator in addition to the proposer reward.

Reviewer: ```suggestion
At each `BeginBlock`, portion of the fees received in the previous block are transferred to the distribution `ModuleAccount` - the account which keeps track of the coin flow for the distribution. The reminder of the fees are allocated to the proposer, community fund and global pool. The block proposer of the previous round (and it's delegators) also receives between 1% and 5% of fee rewards, the reserve community tax is then charged, then the remainder is distributed proportionally by voting power to all bonded validators independent of whether they voted (social distribution). Note the social distribution is applied to proposer validator in addition to the proposer reward.
```

Author: One correction: all the fees are sent to the ModuleAccount, and stay there until they are withdrawn. All the remaining operations are about updating the module-internal bookkeeping. That bookkeeping is used to send out the right amounts at withdraw time. Like a bank which has all its money in a big vault, but has books that keep track of which customer has which balance.

Other than that, I like the bullet style, I'll add that.

> The amount of proposer reward is calculated from pre-commits Tendermint messages in order to incentivize validators to wait and include additional pre-commits in the block. All provision rewards are added to a provision reward pool which validator holds individually (`ValidatorDistribution.ProvisionsRewardPool`).

Reviewer: ```suggestion
The block proposer reward is calculated from Tendermint pre-commit messages in order to incentivize validators to wait and include additional pre-commits in the block. All provision rewards are added to a provision reward pool which validator holds individually (`ValidatorDistribution.ProvisionsRewardPool`).
```

Author: I'm not sure what is meant by 'provision rewards'. I didn't write these paragraphs, they only appear as new text because the file changed name from `03_end_block.md` to `03_begin_block.md`. I'm hesitant to remove the existing text since it may use lingo that is in popular use but that I'm unfamiliar with. I'm happy to remove the paragraphs, or just shorten them down (as suggested below) if they only seem to cause confusion.

Reviewer: Validator takes a provision (commission), eg 5%, for his work. So he is not distributing all rewards - he will firstly take the provision, and then distribute the remaining (eg 95%) to delegators.

## The Distribution Scheme

See [params](07_params.md) for description of parameters.

Let `fees` be the total fees collected in the previous block. All fees are
collected in a specific module account during the block. During `BeginBlock`,
they are sent to the `"distribution"` `ModuleAccount`. No other sending of
tokens occur. Instead, the rewards each account is entitled to are stored, and
withdrawals can be triggered through the messages `FundCommunityPool`,

> `WithdrawValidatorCommission` and `WithdrawDelegatorReward`.

Reviewer: Seams we are repeating the previous paragraph. Suggestion: let's make the previous paragraph shorter and only mention there that the reward allocation happens in the Begin Block for the transactions and proposer of the previous block.

Author: Done! Please have a look again and tell me what you think

### Reward to the Community Pool

> The community pool gets `communitytax * fees`, plus any remaining dust after

Author: This is based on the protocol params, mentioned in the `07_params.md` file. They don't use either snake or camel case. But I don't see why we shouldn't use snake case here, it's clearere.

validators get their rewards, which will always be rounded down to nearest
integer value.

### Reward To the Validators

The proposer will receive a base reward of `fees * baseproposerreward`. In
addition, they will receive a bonus of `fees * bonusproposerreward * P`, where

> `P = (precommits included / total bonded validator power)`. The more precommits

Reviewer: let's be more precise here. On the right hand of the equation we have different units: `total bonded validator power` is measured in staked tokens,  `precommits_included` is measured in number of precommits. I think the precommits must be also scaled based on the bond of the validator who sent the precommit.

Reviewer: Also `total bonded validator power` might be ambiguous: 
* sum of all bonds of all validators (`sum bonds`)
* single validator total bond.

That's why I suggest to use `sum bonds`

Author: I'm not so sure about the term `bonds`, it doesn't really exist anywhere. There is a certain number of validators which are "bonded", meaning they are participating in consensus. But that doesn't mean there is a `bond` in any sense.

You're right about the units. I was hoping that "precommits" would be understandable in context as "the power represented by all included precommits", but clearly that's not the case.

Reviewer: BTW, thte `bonusproposerreward` word should have use proper word separators.

> the propser includes, the larger `P` will be. It can never be larger than `1.00`

Reviewer 2: ```suggestion
the proposer includes, the larger `P` will be. It can never be larger than `1.00`
```

, and will always be larger than `2/3`.

Any remaining fees are distributed among all the bonded validators, including
the proposer, in proportion to their consensus power.

```

> powFrac = validator power / total bonded validator power

Reviewer: ```suggestion
powFrac = validator bond / sum all bonds
```

Author: Not sure "bond" is the right word to use. Validators and stakes are "bonded", but they are not really "bonds".

proposerMul = baseproposerreward + bonusproposerreward * P
voteMul = 1 - communitytax - proposerMul
```

> In total, the proposer will receive `fees * voteMul * powFrac + fees * proposerMul`.

Reviewer: ```suggestion
In total, the proposer will receive `fees  * (voteMul * powFrac + proposerMul)`.
```

> All other validators will receive `fees * voteMul`.

Reviewer: this is wrong - it's too much, because `propose + other_validators + community_tax  * fee  > fee`

Author: Hmm, looks like there's a factor missing, should be `fees * voteMul * powFrac`

### Rewards to Delegators

> Each validators rewards are distributed to its delegators. Note that the
> validator also has a self-delegation.
> 
> The validator sets a commission rate. The commission rate is flexible, but each
> validator sets a max rate which may never be exceeded, and a max daily increase
> which may never be exceeded. This protects delegators against a validator
> suddenly increasing their commission rate, taking all the rewards.

Reviewer 3: ```suggestion
Each validator's rewards are distributed to its delegators. The
the validator also has a self-delegation.

The validator sets a commission rate. The commission rate is flexible, but each
validator sets a maximum rate and a maximum daily increase. These maximums cannot be exceeded and protect delegators from sudden increases of validator commission rates to prevent validators from taking all of the rewards.
```

> The outstanding rewards that the operator is entitled to are stored
> in `ValidatorCurrentRewards`, while the rewards the delegators are entitled to
> are stored in `ValidatorCurrentRewards`.
> The [F1 fee distribution scheme](01_concepts.md) is used to calculate the
> rewards per delegator as they withdraw or update their delegation, and is thus
> not handled in `BeginBlock`.
> 

Reviewer 3: ```suggestion
- The outstanding rewards that the operator is entitled to are stored
in `ValidatorCurrentRewards`. 
- The rewards that the delegators are entitled to
are stored in `ValidatorCurrentRewards`.

The [F1 fee distribution scheme](01_concepts.md) is used to calculate the
rewards per delegator as they withdraw or update their delegation and is not handled with `BeginBlock` operations.

```

Author: Hmm, seems there's a mistake here, they shouldn't both be `ValidatorCurrentRewards`. I'll have a look at what it should be.

### Example Distribution

> As an example, assume the underlying consensus engine selects block proposers in
> proportion to their power relative to the entire bonded power. Further, assume
> all validators are equally good at including pre-commits in their proposed
> blocks. Then we hold `(precommits included) / (total bonded validator power)`
> constant, and then the amortized block reward for the validator is
> simply `( validator power / total bonded power) * (1 - community tax rate)` of
> the total rewards. Consequently, the reward for a single delegator is:

Reviewer 3: ```suggestion
For this example distribution, the underlying consensus engine selects block proposers in
proportion to their power relative to the entire bonded power. 

All validators are equally performant at including pre-commits in their proposed
blocks. Then hold `(precommits included) / (total bonded validator power)`
constant so that the amortized block reward for the validator is `( validator power / total bonded power) * (1 - community tax rate)` of
the total rewards. Consequently, the reward for a single delegator is:
```

<!--
order: 3
-->

# Begin Block

At each `BeginBlock`, all fees received in the previous block are transferred to
the distribution `ModuleAccount` account. When a delegator or validator
withdraws their rewards, they are taken out of the `ModuleAccount`. During begin
block, the different claims on the fees collected are updated as follows:

- The block proposer of the previous round and its delegators receive between 1% and 5% of fee rewards.
- The reserve community tax is charged.
- The remainder is distributed proportionally by voting power to all bonded validators

> To incentivize validators to wait and include additional pre-commits in the block, the block proposer reward is calculated from Tendermint pre-commit messages.

Reviewer 4: Looking at https://github.com/cosmos/cosmos-sdk/blob/master/x/distribution/abci.go#L31-L32

Am I reading this right that when we execute block N, the proposer for block N-1 gets the proposer reward based on the LastCommit included in block N? Which is to say the proposer of block N-1 is getting rewarded based on what the proposer of block N is doing? If this is the case there would appear to be a griefing attack where the proposer of block N just includes the minimal commit set (ie. only 67% votes rather than closer to 100%) such that the proposer of block N-1 will only get the minimum proposer reward. In practice a proposer could just not wait the TimeoutCommit before proposing, since there's technically no incentive to (ie. they'll get paid based on what the next proposer does, not what they do - the previous proposer gets paid based on what they do). Is that right or am I missing something? If it's somehow not a bug, I guess it should probably at least get some extra documentation.

On a related note, do you know of some reason all this logic needs to happen in BeginBlock of block N? Could it instead happen in the EndBlock of block N-1 ?

Reviewer 5: Yeah, these discussions were had so long ago and not all of them were documented unfortunately, but I am 95% sure this is intentional.

See: https://github.com/cosmos/cosmos-sdk/issues/2525#issuecomment-430838701

Reviewer 4: I think #2525 is referring to a different issue in the protocol itself, where validators can withhold votes from particular proposers to grief them. The issue I've identified seems to be an implementation bug, where proposers are getting rewarded based not on what they include in their own block, but on what the next proposer does.

Author: @ebuchman Why do you conclude that the sum of total votes is based on the precommits in the current block? I'm not familiar with what happens in Tendermint, but the votes are read with this line:
```
	for _, voteInfo := range req.LastCommitInfo.GetVotes() {
```
I believe that the `LastCommitInfo` contains info about the previous block. If it doesn't, I agree that it seems like a bug.

From what I gather from the issue @alexanderbez posted, that is more about how a validator can exclude other validators, eventually possibly cause them to get slashed due to liveness issues, but it would be hard to do and would come at some cost to the malicious validator. If, on the other hand, the previous proposer gets bonuses based on the current proposer's included precommits that
a) makes no sense at all, and
b) means that a malicious validator can do this "liveness" attack at 0 cost to themselves.

Can we verify that the LastCommitInfo actually passes the info from the previous block? If it doesn't I'd classify this as a bug, possibly caused by the fact that the distribution used to happen in `EndBlock` and has since been moved to `BeginBlock`.

Reviewer 4: > I believe that the LastCommitInfo contains info about the previous block.

This is true, but those votes are included in the current block - ie. block N includes data called LastCommit, which is the commit for block N-1. The proposer for block N is responsible for putting this commit together and including it in block N, even though it's the commit for block N-1 (that's how Tendermint works and why the field is called LastCommit and not Commit). 

> possibly caused by the fact that the distribution used to happen in EndBlock and has since been moved to BeginBlock.

This seems super important and probably the source of the bug. If this was happening in EndBlock all would be well again :D

Author: So let's see if I have this straight.

Proposer for N-1 puts together a block: they run `BeginBlock`, the transactions, then `EndBlock`. The package it up and send it out to the other validators. This block is signed with some number of precommits, and agreed upon, and the next round starts.

Proposer for N puts together a block. They start with `BeginBlock`. Here they look at all the precommits from the previous round, package them up and include them (how?). The proposer for N-1 gets rewarded based on how many of these precommits are included.

Is that a correct understanding of what you are describing?

It seems the correct behavior is: in BeginBlock, the proposer of *the current block* gets rewarded based on how many precommits of the previous block they included.

> It would still be good to confirm this with a test, for instance have the first proposer make a block and include a commit for 100% of the voting power and then the second proposer makes a block with a commit that contains only 67% of the voting power and see how much the first proposer gets rewarded. It should be the maximum but with the current bug I believe it would end up being the minimum!

Good idea! That's beyond my expertise (and current time available) but who would be able to run a test like that?

> This seems super important and probably the source of the bug. If this was happening in EndBlock all would be well again :D

Actually, if it works the way I think (described above) it shouldn't matter, because `EndBlock` would still get run before the precommits come in, so it would be essentially the same thing. Then it's better to have it in `BeginBlock`, I think, because it doesn't depend on any of the events in the block.

Author: @ebuchman would you mind making this an issue? I think this PR can still be merged (really this piece of documentation seems to reflect the common understanding of what the code does or should do), and I want to make sure that we have an issue where we can keep the discussion in less precarious form.

But please reference this discussion and code location in the issue, so we remind ourselves to update it if necessary.

Reviewer 4: Thanks opened an issue: https://github.com/cosmos/cosmos-sdk/issues/9161

> Proposer for N-1 puts together a block: they run BeginBlock, the transactions, then EndBlock. The package it up and send it out to the other validators. This block is signed with some number of precommits, and agreed upon, and the next round starts.

Note this isn't quite right. Currently a proposer puts together a block (sat N-1), but it is not executed until it is committed (ie. +2/3 precommits for it are seen). However the canonical commit is only determined by the next proposer when they put together block N and include the LastCommit.

> Proposer for N puts together a block. They start with BeginBlock. Here they look at all the precommits from the previous round, package them up and include them (how?). The proposer for N-1 gets rewarded based on how many of these precommits are included.

Again not quite, the proposer doesn't run BeginBlock before proposing. They collect precommits they've seen for block N-1 (there should be at least +2/3, but they'll wait TimeoutCommit to collect more since they're supposed to be rewarded for doing so), and they include those in the proposed block N as the LastCommit. They are supposed to be rewarded based on the size of LastCommit they include but currently the reward for that is going to the previous proposal. 

It seems correct because proposer for block N gets rewarded based on the canonical commit for block N, but the commit for block N is included in block N+1 by the proposer for N+1 so it's actually not correct to reward the proposer of N for the commit of block N, they should be rewarded for the LastCommit contained within block N (which they selected) which is the commit for block N-1. The off-by-one is super confusing :) 

> the proposer of the current block gets rewarded based on how many precommits of the previous block are included.

yes this would be correct behaviour

Author: Okay, still just trying to get my head around the commit process. How can the proposer for N-1 put together a block without actually executing it (BeginBlock + txs + EndBlock)? Or does the proposer for N-1 select a set of txs, bundle them, collect precommits on the bundle, then execute the block and send off the final block to everyone by executing the block? If so, the proposer for N-1 should be able to include the precommits for the current block in BeginBlock.

## The Distribution Scheme

See [params](07_params.md) for description of parameters.

Let `fees` be the total fees collected in the previous block. All fees are
collected in a specific module account during the block. During `BeginBlock`,
they are sent to the `"distribution"` `ModuleAccount`. No other sending of

> tokens occur. Instead, the rewards each account is entitled to are stored, and

Reviewer 3: ```suggestion
tokens occurs. Instead, the rewards each account is entitled to are stored, and
```

> withdrawals can be triggered through the messages `FundCommunityPool`,

Author: To the community pool, yes.

`WithdrawValidatorCommission` and `WithdrawDelegatorReward`.

### Reward to the Community Pool

The community pool gets `community_tax * fees`, plus any remaining dust after
validators get their rewards that are always rounded down to the nearest
integer value.

### Reward To the Validators

The proposer receives a base reward of `fees * baseproposerreward` and a bonus
of `fees * bonusproposerreward * P`, where `P = (total power of validators with
included precommits / total bonded validator power)`. The more precommits the
proposer includes, the larger `P` is. `P` can never be larger than `1.00` (since

> ony bonded validators can supply valid precommits) and is always be larger than

Reviewer 3: hi @ebuchman wdyt about following the advice of using present tense, even in specs? I propose https://developers.google.com/style/tense

> ony bonded validators can supply valid precommits) and is always be larger than

Reviewer 4: Oh sure that sounds great, sentence would still needs fixing though

`2/3`.

Any remaining fees are distributed among all the bonded validators, including
the proposer, in proportion to their consensus power.

```
powFrac = validator power / total bonded validator power
proposerMul = baseproposerreward + bonusproposerreward * P
voteMul = 1 - communitytax - proposerMul
```

In total, the proposer receives `fees  * (voteMul * powFrac + proposerMul)`.
All other validators receive `fees * voteMul * powFrac`.

### Rewards to Delegators

Each validator's rewards are distributed to its delegators. The validator also

> has a self-delegation, which is treated like a regular delegation in

Reviewer 3: ```suggestion
has a self-delegation that is treated like a regular delegation in
```
